proposals/gc/Overview.md

```diff
@@ -1 +1,728 @@
+```
+This extends to nested and recursive types:
+```
+(type $t1 (struct (type $pt) (ptr $t2)))
+(type $t2 (struct (type $pt) (ptr $t1)))  ;; t2 = t1
+(type $u (struct (type $vec) (ptr $u)))   ;; u = t1 = t2
+```
+Note: This is the standard definition of recursive structural equivalence for "equi-recursive" types.
+Checking it is computationally equivalent to checking whether two FSAs are equivalent, i.e., it is a non-trivial algorithm (even though most practical cases will be trivial).
+This may be a problem, in which case we need to fall back to a more restrictive definition, although it is unclear what exactly that would be.
+
+
+### Subtyping
+
+Subtyping is designed to be _non-coercive_, i.e., never requires any underlying value conversion.
+
+The subtyping relation is the reflexive transitive closure of a few basic rules:
+
+1. The `anyref` type is a supertype of every reference type (top reference type).
+2. The ` anyfunc` type is a supertype of every function type.
+3. A structure type is a supertype of another structure type if its field list is a prefix of the other (width subtyping).
+4. A structure type is a supertype of another structure type if they have the same fields and for each field type:
+   - The field is mutable in both types and the storage types are the same.
+   - The field is immutable in both types and their storage types are in (covariant) subtype relation (depth subtyping).
+5. An array type is a supertype of another array type if:
+   - Both element types are mutable and the storage types are the same.
+   - Both element types are immutable and their storage types are in
+(covariant) subtype relation (depth subtyping).
+6. A function type is a supertype of another function type if they have the same number of parameters and results, and:
+   - For each parameter, the supertype's parameter type is a subtype of the subtype's parameter type (contravariance).
+   - For each result, the supertype's parameter type is a supertype of the subtype's parameter type (covariance).
+
+Note: Like [type equivalence](#type-equivalence), subtyping is *structural*.
+The above is the standard (co-inductive) definition, which is the most general definition that is sound.
+Checking it is computationally equivalent to checking whether one FSA recognises a sublanguage of another FSA, i.e., it is a non-trivial algorithm (even though most practical cases will be trivial).
+Like with type equivalence, this may be a problem, in which case a more restrictive definition might be needed.
+
+Subtyping could be relaxed such that mutable fields/elements could be subtypes of immutable ones.
+That would simplify creation of immutable objects, by first creating them as mutable, initialize them, and then cast away their constness.
+On the other hand, it means that immutable fields can still change, preventing various access optimizations.
+(Another alternative would be a three-state mutability algebra.)
+
```

Comment:
So, types are organized in a tree, which is insufficient for Java, where types form DAG. How are interfaces supposed to be implemented?